src/services/game-logic/GameEngine.ts

```diff
@@ -10,8 +12,24 @@ class GameEngine {
 
     getLegalMoves = (square: Square): Square[] => {
         const piece = this.board.getPiece(square);
-        return piece?.getPossibleMoves() ?? [];
+        let pieceMoves = piece?.getPossibleMoves() ?? [];
+        return pieceMoves
+            .filter(this.isOnBoard)
+            .filter((square) => !this.isPawnAndBlocked(square, piece))
+            .filter((square) => !this.isOccupiedBySameColor(square, piece));
     };
+
+    private isOnBoard = (square: Square): Boolean => {
+        // here or in knight.ts (for now knights can get outside the board)
+        return square.row >= 0 && square.row < 8 && square.column >= 0 && square.column < 8;
+    };
+
+    private isPawnAndBlocked = (square: Square, piece: Piece | null): Boolean => {
```

Comment:
Jeżeli to jest inline z tym co robi Gosia to może zostać, ja w weekend postaram się usiąść pomyśleć i rozpisać taska by to udoskonalić. 

Mój problem z taką implementacją jest że mamy klase Pawn która ma z założenia odpowiadać za pionka i wszystko co z nim związane, szczegółowe pomysły jak ugryźć temat w PRce Gosi